### ISSUE TITLE
Issue with dataloader using pin_memory = True

### ISSUE THREAD

**Issue by ankur6ue** [pytorch/pytorch#16901]
## Issue description
Hello, I'm seeing an odd issue with using the pin_memory = true flag with the dataloader. I'm measuring the time taken to transfer data from the host RAM to GPU memory as follows:

        transfer_time_start = time.time()
        input = input.cuda(args.gpu, non_blocking=False)
        target = target.cuda(args.gpu, non_blocking=False)
        torch.cuda.synchronize()
        transfer_time.update(time.time()-transfer_time_start)

with pin_memory = True in the dataloader, this gives me a transfer time of 0.03 sec, which for a batch size of 256, translates into 256*224*224*3*4/0.03 = 5.1GB, which is a bit low for my CPU-GPU interconnect (x16, PCIe3) which should deliver ~12GB. 

I then tried calling pin_memory() manually on the tensor returned by the enumerate call, as shown below:

for i, (input, target) in enumerate(train_loader):
        input = input.pin_memory()
        # measure data loading time
        data_time.update(time.time() - end)
        transfer_time_start = time.time()
        input = input.cuda(args.gpu, non_blocking=False)
        target = target.cuda(args.gpu, non_blocking=False)
        torch.cuda.synchronize()
        transfer_time.update(time.time()-transfer_time_start)

Now the transfer time dropped to 0.014, which translates to ~11GB, which is as expected. Anyone has any ideas why setting pin_memory = True in the data loader may not return a tensor already in pinned memory?

Also attached below are two plots showing the transfer time (green plot) from host memory to the GPU.
This plot shows the transfer time when I call pin_memory manually
![image|690x208](upload://2jdPCZgJNcjw5ZK4YqflrDTq6gQ.png) 
You can see that the transfer time stays consistently low.

Whereas this one shows the transfer time without calling pin_memory manually. Now the transfer time is highly variable and averages to around 0.03 sec
![image|624x247](upload://T0uysPWY4omqgGG0H8V3gYyMGn.png)

If I insert a sleep of 50 ms after the enumerate call, I again obtain a nice, low transfer time. This indicates that the data loader simply needs a bit more time to finish the pinned_memory transfer? But inserting the sleep shouldn't be necessary as the dataloader shouldn't return from the enumerate call if the transfer is not complete yet? 

## System Info
Please copy and paste the output from our
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 387.26
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.2

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] torch==1.0.0
[pip] torchvision==0.2.1
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl_fft                   1.0.10           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] pytorch                   1.0.0           py3.7_cuda9.0.176_cudnn7.4.1_1    pytorch
[conda] torchvision               0.2.1                      py_2    pytorch

**Comment by SsnL** [comment]
Try inserting a `cuda.synchronize()` before recording `transfer_time_start`

**Comment by ankur6ue** [comment]
I tried that but it didn't make a difference. However I'll do it again on monday to be sure. I did record the time it takes to execute batch = pin_memory_batch(batch) in dataloader and it is a reasonable, non zero amount of time, indicating that this is not an asynchronous call.. 

Any other ideas?

**Comment by ailzhang** [comment]
@ankur6ue  Please let us know what you see after applying @SsnL 's suggestion. We should follow up on this if the issue still persists.

**Comment by SsnL** [comment]
That’s interesting. Could you try coming up with a script I can use to reproduce and investigate this? Thanks!

**Comment by ankur6ue** [comment]
yes sure. My script is basically the imagenet training script with some plotting code added. I'll attach the script and reproduction instructions to this thread tomorrow. Thanks very much for taking a look.

**Comment by ankur6ue** [comment]
The code is attached. It is the imagenet training script with timing code added. Please note the cuda.synchronize() calls at the end of host to device data transfer and data processing. Those are needed to measure the timing accurately. Without the synchronize calls, the transfer time is blended into the processing time. 
[imagenet_data_parallel.py.zip](https://github.com/pytorch/pytorch/files/2851753/imagenet_data_parallel.py.zip)

The code can be run as follows:
PS: please replace the /home/ankur/dev/tiny-imagenet-200 with appropriate directory on your system.
 # Without pin_memory in the main thread:
python3 imagenet_data_parallel.py /home/ankur/dev/tiny-imagenet-200 --batch-size 256 --workers 4 --arch resnet18 --gpu 0

This should result in a transfer time of ~0.035, which is slow. 
# With pin_memory in the main thread:
uncomment line# 296 (input = input.pin_memory())
and run the code again. Now the transfer time drops to ~0.015, which is as expected. 

I've already provided system info above. Thanks again for taking a look.

**Comment by ankur6ue** [comment]
I forgot to add - SsnL's suggestion didn't make a difference.

**Comment by ankur6ue** [comment]
Any updates on this? 

I did some experiments on how multi-threading can be used to perform unpinned-pinned memory transfer concurrently with host-to-device data transfer and data processing on device, which is the approach taken by the dataloader. Here's a summary:

http://www.telesens.co/2019/02/16/efficient-data-transfer-from-paged-memory-to-gpu-using-multi-threading/

So the technique appears to work, at least in my simple example. Would be good to understand whats going on with the dataloader..

**Comment by xmyqsh** [comment]
@ankur6ue awesome summary!
